compiler/modules/front/expr.bal

```diff
@@ -533,11 +551,110 @@ function codeGenArithmeticBinaryExpr(ExprContext cx, bir:BasicBlock bb, bir:Arit
     return { result, block: bb };
 }
 
-function codeGenLogicalBinaryExpr(ExprContext cx, bir:BasicBlock bb, s:BinaryLogicalOp op, Position pos, bir:BooleanOperand lhs, bir:BooleanOperand rhs) returns CodeGenError|ExprEffect {
-    if lhs is boolean && rhs is boolean {
-        return { result: logicalEval(op, lhs, rhs), block: bb };
+function codeGenLogicalBinaryExpr(ExprContext cx, bir:BasicBlock bb, s:BinaryLogicalOp op, Position pos, s:Expr left, s:Expr right) returns CodeGenError|ExprEffect {
+    var { result: lhs, block: block1, narrowing: lhsNarrowing } = check codeGenExprForBoolean(cx, bb, left);
+    if lhs is boolean {
+        boolean shortCircuit = op == "||";
+        if lhs == shortCircuit {
+            return <ExprEffect>{ result: shortCircuit, block: bb };
+        }
+        else {
+            return codeGenExprForBoolean(cx, block1, right);
+        }
+    }
+    final bir:Register result = lhs; // reuse lhs register as result
```

Comment:
Make a new one.

Reply:
Fixed, but I had to add an additional BasicBlock.